### PR TITLE
Add Theatron to Architecture_oM and Humans_oM

### PR DESCRIPTION
**Pull request by rolyhudson**
Theatron is a set of Architecture objects for defining the geometry of stadia, theatre, cinemas, lecture halls and auditoria. It includes Human objects related to users of these spaces: Spectators, Audience, Eye, ViewCones and analytic methods for evaluating view quality and ergonomics

### NOTE: Depends on 
A fork of BHoM_Engine with object definitions in Architecture_Engine and Humans_Engine
pull request in BHoM_Engine: https://github.com/BHoM/BHoM_Engine/pull/1107 
### Test files
Three zipped grasshopper files that demonstrate three key use case scenarios
[theatron-GH_Tests.zip](https://github.com/BHoM/BHoM_Engine/files/3418742/theatron-GH_Tests.zip)

### Additional comments
This is first time pull request to BHoM so its wrong, too complicated or missing something please let me know

### CODE REVIEW

**Comment by rolyhudson** [comment]
Hi @al-fisher :sunglasses: thanks for reviewing. I have skimmed your thoughts all really helpful stuff :thumbsup: :thumbsup:. Just starting to go through your comments one by one and will send further thoughts.

File: Architecture_oM/Theatron/Elements/ActivityArea.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+        public Polyline PlayingArea { get; set; } = new Polyline();
+
+        public Point ActivityFocalPoint { get; set; } = new Point();
+
+        public double Width { get; set; } = 0.0;
+
+        public double Length { get; set; } = 0.0;
+
+        /***************************************************/
```
[the remark this answers is not in the document]
Comment (rolyhudson) [reply Architecture_oM/Theatron/Elements/ActivityArea.cs]: @al-fisher Width and Length properties are obsolete now as they relate only to the football / rectangular pitch stadia use case which is quite a special case. I'll check this remove them an see if there are other implications. I'll also take a look at the ShapeProfiles you mention. I think the Polyline may allow greatest flexibility. In the case of an analysis of a curved cinema screen this is input as a 3d Polyline for the ActivityArea.

File: Architecture_oM/Theatron/Elements/SeatingBlock.cs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+        public ProfileOrigin Vomitory { get; set; } = new ProfileOrigin();
+
+        public ProfileOrigin End { get; set; } = new ProfileOrigin();
+
+        public List<TierProfile> Sections { get; set; } = new List<TierProfile>();
+
+        public Mesh Floor { get; set; } = new Mesh();
+
+        public Audience Audience { get; set; } = new Audience();
```
[the remark this answers is not in the document]
Comment (rolyhudson) [reply Architecture_oM/Theatron/Elements/SeatingBlock.cs]: :thumbsup:

File: Humans_oM/ViewQuality/AvalueSettings.cs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+        public ViewConeEnum ConeType { get; set; } = ViewConeEnum.Undefined;
+
+        public bool CalculateOcclusion { get; set; } =  false;
+
+        public double EyeFrameDist { get; set; } = 0.1;
+
+        public double ForeheadSize { get; set; } = 0.120;
+
+        public double NearHeadRange { get; set; } = 0.100;//dist from viewplane centre for nearest heads
```
[the remark this answers is not in the document]
Comment (rolyhudson) [reply Humans_oM/ViewQuality/AvalueSettings.cs]: @al-fisher Yes that is much needed across the Theatron_oM and related HUmans_oM

File: Humans_oM/ViewQuality/Spectator.cs [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+namespace BH.oM.Humans.ViewQuality
+{
+    public class Spectator : BHoMObject, IHumanRole
+    {
+        /***************************************************/
+        /**** Properties                                ****/
+        /***************************************************/
+
+        public BodyParts.Eye Eye { get; set; } = new  BodyParts.Eye();
```
[the remark this answers is not in the document]
Comment (rolyhudson) [reply Humans_oM/ViewQuality/Spectator.cs]: in the land of the blind the one eyed man is king :joy: :joy: :joy:. EyePair plus a ref point is a good idea

[the remark this answers is not in the document]
Comment (al-fisher) [reply Humans_oM/ViewQuality/Spectator.cs]: Cool - I am really liking the idea of a `PairOfEyes` object now actually! 👀 
Think this approach can be really useful for a `PairOfEars` too in the future for our acoustical friends and binocular sounds. 👂👂

**Comment by rolyhudson** [comment]
> I was also looking at the Parameters objects you have created and how they could potentially be broken into either smaller identifiable subcategories. Or assigned as Properties directly to the number of host objects rather than flat list of variables. The idea being to make more modular and increase flexibility of workflow and model generation. 

This is a good point @al-fisher especially with the ``` Theatron ``` objects in ``` Architecture_oM```  and ``` ViewQuality```  in ``` Humans_oM``` . It makes sense to pull out parameters related to eye positions -seated, standing in wheelchair and put those into a  ```EyePositionParameters``` object in ``` ViewQuality```  in ``` Humans_oM```. Other logical modules that I'm splitting out are those related to vomitories and those related to SuperRisers

File: Humans_oM/ViewQuality/Spectator.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (rolyhudson) [reply Humans_oM/ViewQuality/Spectator.cs]: Looking at this I think the Spectator should have a ```Head``` property and ```Head``` should have the ```PairOfEyes``` property and a 2d Polyline ```HeadOutline```
I'm using ```HeadOutline``` in evaluation of occluded views in Avalue so i think this hierarchy makes sense

File: Architecture_oM/Theatron/Parameters/ProfileParameters.cs [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        public double SuperRiserEyePositionX { get; set; } = 0.0;
+
+        public double SuperRiserEyePositionZ { get; set; } = 0.0;
+
+        public double StandingEyePositionZ { get; set; } = 0.0;
+
+        public double StandingEyePositionX { get; set; } = 0.0;
+
+        public double RiserHeightRounding { get; set; } = 0.0;
+
```
Comment (rolyhudson) [inline Architecture_oM/Theatron/Parameters/ProfileParameters.cs]: Broken these parameters into modules
EyePositionParameters (in Humans_oM), VomitoryParameters and SuperRiserParameters

File: Humans_oM/ViewQuality/Spectator.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (rolyhudson) [reply Humans_oM/ViewQuality/Spectator.cs]: > Looking at this I think the Spectator should have a `Head` property and `Head` should have the `PairOfEyes` property and a 2d Polyline `HeadOutline`
> I'm using `HeadOutline` in evaluation of occluded views in Avalue so i think this hierarchy makes sense

Made this change and fixed Engine related issues

**Comment by rolyhudson** [comment]
I've addressed all these suggested changes and committed. Still to add the classic stadium parameters to Datasets. Also updated the GH demos:
I'm not quite sure whats next in the process but let me know if action is needed from me. 👍 👍 👂 👀 👂 

[theatron GH demos.zip](https://github.com/BHoM/BHoM/files/3458810/theatron.GH.demos.zip)

**Pull request merged**